integration-tests/ahm/src/mock.rs

```diff
@@ -184,24 +186,36 @@ pub fn enqueue_ump(msgs: Vec<UpwardMessage>) {
 	}
 }
 
-#[cfg(feature = "ahm-polkadot")] // XCM V3
+#[cfg(feature = "ahm-polkadot")] // XCM V3 or V4
 fn sanity_check_xcm<Call: Decode>(msg: &[u8]) {
 	let xcm = xcm::VersionedXcm::<Call>::decode(&mut &msg[..]).expect("Must decode DMP XCM");
-	let xcm = match xcm {
-		VersionedXcm::V3(inner) => inner.0,
+	match xcm {
+		VersionedXcm::V3(inner) =>
+			for instruction in inner.0 {
+				match instruction {
+					xcm::v3::Instruction::Transact { call, .. } => {
+						// Interesting part here: ensure that the receiving runtime can decode the
+						// call
+						let call: Call = Decode::decode(&mut &call.into_encoded()[..])
+							.expect("Must decode DMP XCM call");
+					},
+					_ => (), // Fine, we only check Transacts
+				}
+			},
+		VersionedXcm::V4(inner) =>
```

Comment:
We can also merge the Westend and Polkadot functions now that it is in the same function already.

Reply:
we need to merge master to dev brach first for that